Libraries to collections

martenson opened the pull request:
this is tied to the client API that will be probably made obsolete by #5038 so I will continue with this after I review that

ping @jmchilton 

edit: this is ready for review now

martenson committed:
convert to arrow functions, eliminate self and that

martenson committed:
remove styles for the removed old library interface

martenson committed:
wip on importing from library to history as a collection

+ various es6 additions; cleanup; refactoring

I tried to separate it into more commits but it kept growing too fast, sorry

jmchilton commented on static/style/blue/base.css:
```diff
@@ -1,4 +1,8 @@
 /*! normalize.css v2.1.0 | MIT License | git.io/normalize */article,aside,details,figcaption,figure,footer,header,hgroup,main,nav,section,summary{display:block}audio,canvas,video{display:inline-block}audio:not([controls]){display:none;height:0}[hidden]{display:none}html{font-family:sans-serif;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%}body{margin:0}a:focus{outline:thin dotted}a:activ ...
```
Bad merge here

martenson replied:
This is just a static file though, the whole PR is not `make client`'d just this file got there for some reason. Should be fine after build I think.

jmchilton replied:
Yup - good points - I didn't mean to nit pick.

martenson committed:
implement support for paired and list:paired collections

martenson committed:
finish the list:paired modal invocation

martenson commented:
ready for fist review I think - ping @jmchilton

jmchilton commented on client/galaxy/scripts/mvc/collection/list-collection-creator.js:
```diff
@@ -278,9 +278,9 @@ var ListCollectionCreator = Backbone.View.extend(BASE_MVC.LoggableMixin)
-            }
+            // if (element.history_content_type !== "dataset") {
+            //     return _l("is not a dataset");
+            // }
```
Ha! That was simple it seems - very nice!

martenson replied:
I improved that so it respects data coming from datalibrary

martenson commented:
![screenshot 2017-12-18 12 04 44](https://user-images.githubusercontent.com/1814954/34123936-27559b0c-e3ff-11e7-96d2-0773a1268a68.png)
![screenshot 2017-12-18 12 04 53](https://user-images.githubusercontent.com/1814954/34123937-276501b4-e3ff-11e7-8df6-bbcc6366329f.png)

martenson committed:
re enable and enhance the collection validator

jmchilton commented on client/galaxy/scripts/mvc/collection/list-collection-creator.js:
```diff
@@ -278,9 +278,9 @@ var ListCollectionCreator = Backbone.View.extend(BASE_MVC.LoggableMixin)
-            // }
+            if (element.history_content_type !== "dataset" && element.type !== "file") {
+                return _l("is not a dataset or a library file");
+            }
```
Well I liked it better before you did this - there is no reason in my mind for this extra check.

martenson replied:
Really? I mean the check is kinda silly indeed, but I thought it serves a purpose of disallowing collections in?

jmchilton replied:
But collections should be allowed in ideally - that used to be allowed - but good point the current version of these wouldn't work with collections.

martenson replied:
Yeah, they should be allowed in, but this is not what is stopping them as you point out.

martenson commented on client/galaxy/scripts/mvc/library/library-foldertoolbar-view.js:
```diff
@@ -1238,6 +1195,175 @@ var FolderToolbarView = Backbone.View.extend({
+    },
+
+    createHDCA: function(elementIdentifiers, collectionType, name, hideSourceItems, history_id, options) {
+        let hdca = new HDCA_MODEL.HistoryDatasetCollection({
```
it would be nice to refactor this together with how the collection builder modals are invoked, but it does not seem too inelegant

martenson committed:
refactor modal to bind closing event at different time

this allows chaining modals after each other with closing events working

jmchilton commented on client/galaxy/scripts/mvc/library/library-foldertoolbar-view.js:
```diff
@@ -1238,6 +1195,175 @@ var FolderToolbarView = Backbone.View.extend({
+                ];
+                return this.createHDCA(elements, this.collectionType, name, hideSourceItems, history_id);
+            };
+            LIST_CREATOR.collectionCreatorModal(
```
PAIR_CREATOR.pairCollectionCreatorModal right?

martenson replied:
The function you mention is just a wrapper over `LIST_CREATOR.collectionCreatorModal` that adds a title which I do not want to use. I can refactor it if you deem it a better approach.

jmchilton replied:
Should we just change the title everywhere then to the one you want - yours looks better? If yes, can we eliminate the other entries points all together?

Also, if the LIST_CREATOR has a good general purpose collection creator method we want to use to create pairs - can we move it into base-creator.js to make it more clear its function?

I know these aren't your problems - I'm happy to clean this up in a follow up PR if you wish. I'd just like to better understand the collection creator entry points.

martenson replied:
>I'd just like to better understand the collection creator entry points.

you and me both 🤣

martenson replied:
@jmchilton The modal titles here are only different in the fact they include name of the history you are creating the collections in - which is only relevant for the datalib use case - when you are invoking the collection modals from history it is obvious which one you are creating it in.

jmchilton commented on client/galaxy/scripts/mvc/library/library-foldertoolbar-view.js:
```diff
@@ -1238,6 +1195,175 @@ var FolderToolbarView = Backbone.View.extend({
+            };
+            LIST_CREATOR.collectionCreatorModal(
+            collection_elements,
+            {creationFn: creationFn, title: modal_title},
```
I think in each case this option array should have ``defaultHideSourceItems: true`` enabled. It'd be better to not see the original HDAs in the history by default. It makes even moderately large collections difficult to work with.

martenson replied:
I suspect this is overruled in the creator modal itself (which has its own checkbox for this, but I will check)

martenson replied:
you are right, the modal respects this and I adjusted my code to take advantage of that and hide the source items by default

jmchilton committed:
Improvements to library testing.

- Add tests for creating a collection from a library.
- Add more class names to library GUI to improve selectors for tests.
- Automatically select whichever radio box makes sense in library collection builder dialog based on whether there is a selection or not.

martenson committed:
hide source items by default in all library-collection creators

martenson committed:
Merge pull request #26 from jmchilton/lib-to-colllection

Improvements to library testing.

martenson committed:
fix the modal backdrop listener to respect the config

martenson committed:
fix the selenium test that expects the class to be here

jmchilton commented:
Thanks for the Selenium fix - sorry about that.

jmchilton committed:
Fix more selenium library tests.

jmchilton committed:
Merge remote-tracking branch 'jmchilton/dev' into lib-to-colllection

jmchilton committed:
Rebuild client.

jmchilton commented:
Awesome thanks for the great new feature @martenson!
